### PR TITLE
add TSConstructorType

### PR DESCRIPTION
**Pull request by despairblue**

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -1516,7 +1516,10 @@ function genericPrintNoParens(path, options, print, args) {
           // TypeScript should not have a colon in TSFirstTypeNode nodes
           // `a is number`
           !(path.getParentNode().type === "TypeAnnotation" &&
```
Comment (vjeux) [inline src/printer.js]: can you group those to in a single condition
```
!(path.getParentNode().type === "TypeAnnotation" &&
  (path.getParentNode().typeAnnotation.type === "TSConstructorType" ||
    path.getParentNode().typeAnnotation.type === "TSConstructorType"))
```

File: src/printer.js [hunk 2]
```diff
@@ -1963,6 +1966,15 @@ function genericPrintNoParens(path, options, print, args) {
       return concat([path.call(print, "expression"), "!"]);
     case "TSThisType":
       return "this";
+    case "TSConstructorType":
+      return concat([
+        "new",
+        "(",
```
Comment (vjeux) [inline src/printer.js]: you don't need to make two strings for this, you can do "new(" directly

File: src/printer.js [hunk 3]
```diff
@@ -1963,6 +1966,15 @@ function genericPrintNoParens(path, options, print, args) {
...
+        join(", ", path.map(print, "parameters")),
+        ")",
+        " => ",
```
Comment (vjeux) [inline src/printer.js]: same here ") => "

**Comment by vjeux** [comment]
Yay!

**Comment by despairblue** [comment]
@vjeux done :)

Also regarding the `TypeAnnotation`, the more corner cases I hit the more I think that it shouldn't be the `TypeAnnotation` that decides if a colon goes there, but rather the parent. :thinking:

**Comment by vjeux** [comment]
Feel free to refactor it to be the parent :)

**Pull request merged**